rust/main/chains/hyperlane-sealevel/src/mailbox.rs

```diff
@@ -733,20 +687,178 @@ impl SealevelMailboxIndexer {
+        )
+        .ok_or_else(|| {
+            ChainCommunicationError::from_other_str(
+                "Could not find program address for unique message pubkey",
+            )
+        })?;
+        Ok(expected_pubkey)
+    }
+
+    async fn get_delivered_message_with_nonce(
+        &self,
+        nonce: u32,
+    ) -> ChainResult<(Indexed<H256>, LogMeta)> {
+        let discriminator = hyperlane_sealevel_mailbox::accounts::PROCESSED_MESSAGE_DISCRIMINATOR;
+        let offset = 1 + 8 + 8; // the offset to get the `message_id` field
+        let length = 32;
+        let accounts = self
+            .search_accounts_by_discriminator(&discriminator, nonce, offset, length)
+            .await?;
+
+        debug!(account_len = ?accounts.len(), "Found accounts with processed message discriminator");
+
+        let valid_message_storage_pda_pubkey = self
+            .search_and_validate_account(accounts, |account| {
+                self.delivered_message_account(&account)
+            })?;
+
+        // Now that we have the valid delivered message storage PDA pubkey,
+        // we can get the full account data.
+        let account = self
+            .rpc()
+            .get_account_with_finalized_commitment(&valid_message_storage_pda_pubkey)
+            .await?;
+        let delivered_message_account = ProcessedMessageAccount::fetch(&mut account.data.as_ref())
+            .map_err(ChainCommunicationError::from_other)?
+            .into_inner();
+        let message_id = delivered_message_account.message_id;
+
+        Ok((
+            message_id.into(),
+            LogMeta {
+                address: self.mailbox.program_id.to_bytes().into(),
+                block_number: delivered_message_account.slot,
+                // TODO: get these when building out scraper support.
+                // It's inconvenient to get these :|
                 block_hash: H256::zero(),
                 transaction_id: H512::zero(),
                 transaction_index: 0,
                 log_index: U256::zero(),
             },
         ))
     }
+
+    fn delivered_message_account(&self, account: &Account) -> ChainResult<Pubkey> {
+        let message_id = H256::from_slice(&account.data);
+        let (expected_pubkey, _bump) = Pubkey::try_find_program_address(
+            mailbox_processed_message_pda_seeds!(message_id),
+            &self.mailbox.program_id,
+        )
+        .ok_or_else(|| {
+            ChainCommunicationError::from_other_str("Could not find program address for message id")
+        })?;
+        Ok(expected_pubkey)
+    }
+
+    fn search_and_validate_account<F>(
+        &self,
+        accounts: Vec<(Pubkey, Account)>,
+        message_account: F,
+    ) -> ChainResult<Pubkey>
+    where
+        F: Fn(&Account) -> ChainResult<Pubkey>,
+    {
+        let mut valid_storage_pda_pubkey = Option::<Pubkey>::None;
+
+        for (pubkey, account) in accounts {
+            let expected_pubkey = message_account(&account)?;
```

Comment:
is it guaranteed that accounts returned from `search_accounts_by_discriminator` exist as message PDAs? Just wondering if someone creates a fake account by discrimantor that then doesn't have an associated `message_account`. Would that propagate an error here and prevent the actually valid account from being checked?

if that is the case, maybe it'd be better to log an error rather than propagate here (and you could also use an iterator with `.find_map` instead of a for loop here)

Reply:
When we create message PDA account, we check that such account was not initialize. Once we created this account, nobody can create another one.

If someone has created such account beforehand, message dispatch will fail, I suppose.